expression-compiler/src/main/scala-3/dotty/tools/dotc/evaluation/InsertExtracted.scala

```diff
@@ -0,0 +1,64 @@
+package dotty.tools.dotc.evaluation
+
+import dotty.tools.dotc.EvaluationContext
+import dotty.tools.dotc.ast.tpd._
+import dotty.tools.dotc.core.Contexts._
+import dotty.tools.dotc.core.Symbols.TermSymbol
+import dotty.tools.dotc.transform.MegaPhase.MiniPhase
+import dotty.tools.dotc.transform.SymUtils.isField
+
+class InsertExtracted(using evalCtx: EvaluationContext) extends MiniPhase:
+  thisPhase =>
+  override def phaseName: String = InsertExtracted.name
+
+  override def transformDefDef(tree: DefDef)(using Context): Tree =
+    if tree.name.toString == "evaluate" then
+      evalCtx.evaluateMethod = tree.symbol
+      val transformedExpression = ExpressionTransformer().transform(
+        evalCtx.expressionValDef
+      )
+      val transformedNestedMethods = evalCtx.nestedMethods.values
+        .map(nestedMethod => ExpressionTransformer().transform(nestedMethod))
+        .toList
+      val expressionBlock =
+        Block(
+          List(transformedExpression),
+          evalCtx.expressionIdent
+        )
+      DefDef(
+        tree.symbol.asInstanceOf[TermSymbol],
+        List(),
+        tree.tpt.tpe,
+        Block(transformedNestedMethods, expressionBlock)
+      )
+    else super.transformDefDef(tree)
+
+  class ExpressionTransformer extends TreeMap:
```

Comment:
can this be an object?

```suggestion
  object ExpressionTransformer extends TreeMap:
```